Lib/site.py

```diff
@@ -426,62 +426,88 @@ def sethelper():
+        # http://bugs.python.org/issue5845#msg198636
+        history = os.path.join(os.path.expanduser('~'),
+                               '.python_history')
         try:
-            readline.read_init_file()
+            readline.read_history_file(history)
         except OSError:
-            # An OSError here could have many causes, but the most likely one
-            # is that there's no .inputrc file (or .editrc file in the case of
-            # Mac OS X + libedit) in the expected location.  In that case, we
-            # want to ignore the exception.
             pass
 
-        if readline.get_current_history_length() == 0:
-            # If no history was loaded, default to .python_history.
-            # The guard is necessary to avoid doubling history size at
-            # each interpreter exit when readline was already configured
-            # through a PYTHONSTARTUP hook, see:
-            # http://bugs.python.org/issue5845#msg198636
-            history = os.path.join(os.path.expanduser('~'),
-                                   '.python_history')
+        def write_history():
             try:
-                readline.read_history_file(history)
+                readline.write_history_file(history)
             except OSError:
+                # bpo-19891, bpo-41193: Home directory does not exist
+                # or is not writable, or the filesystem is read-only.
                 pass
 
-            def write_history():
-                try:
-                    readline.write_history_file(history)
-                except OSError:
-                    # bpo-19891, bpo-41193: Home directory does not exist
-                    # or is not writable, or the filesystem is read-only.
-                    pass
+        atexit.register(write_history)
 
-            atexit.register(write_history)
+def _register_detect_pip_usage_in_repl():
+    old_excepthook = sys.excepthook
+
+    def detect_pip_usage_in_repl(typ, value, traceback):
+        if typ is SyntaxError and (
+            "pip install" in value.text or "pip3 install" in value.text
+        ):
+            value.add_note(
+                "The Python package manager (pip) can only be used"
+                " from outside of Python.\n"
+                "Please try the `pip` command in a"
+                " separate terminal or command prompt."
+            )
+
+        old_excepthook(typ, value, traceback)
+
+    detect_pip_usage_in_repl.__wrapped__ = old_excepthook
+    sys.excepthook = detect_pip_usage_in_repl
+
+
+def _set_interactive_hook():
+    """Register a sys.__interactivehook__ to:
+        - Enable default readline configuration on interactive prompts.
+        - Register an excepthook to detect pip usage in the REPL.
+    """
+    def interactivehook():
+        enablerlcompleter()
+        _register_detect_pip_usage_in_repl()
+
+    sys.__interactivehook__ = interactivehook
 
-    sys.__interactivehook__ = register_readline
```

Comment:
I think this change to the way `enablerlcompleter` works is still concerning.

Reviewing https://github.com/python/cpython/pull/8536/commits/0e6d9e29bbdf7f64c76dcdbe9af6eef226ae5399 I think the public API to preserve would be:

1. New top level function that actually does the registration is called `_register_readline` (as @tomviner originally had it)
2. Reduced `enablerlcompleter` implementation just does `sys.__interactive_hook__ = _register_readline`
3. `_set_interactive_hook` call `_register_readline` directly, bypassing `enablerlcompleter` (since it is setting its own hook)

`enablerlcompleter` would never be called by the standard library (outside the test suite), it would purely be a backwards compatibility API accounting for the fact that it was never marked as private in the interactive help.